includes/class-newspack-popups-data-api.php

```diff
@@ -0,0 +1,176 @@
+			$custom_settings[] = sprintf( __( 'max %d times', 'newspack-popups' ), $popup['options']['frequency_max'] );
+
+			// Translators: %s is the time period for when the prompt can be displayed again after the max number of displays.
+			$custom_settings[] = sprintf( __( 'resetting every %s', 'newspack-popups' ), $popup['options']['frequency_reset'] );
+		}
+
+		return implode( ',', $custom_settings );
+	}
+
+	/**
+	 * Extract the relevant data from a popup.
+	 *
+	 * This method is used by the Newspack Data Events API.
+	 *
+	 * @param int|array $popup The popup ID or object.
+	 * @return array
+	 */
+	public static function get_popup_metadata( $popup ) {
+		if ( is_numeric( $popup ) ) {
+			$popup = Newspack_Popups_Model::retrieve_popup_by_id( $popup );
+		}
+		$data = [];
+		if ( ! $popup ) {
+			return $data;
+		}
+
+		$data['prompt_id']    = $popup['id'];
+		$data['prompt_title'] = $popup['title'];
+
+		if ( isset( $popup['options'] ) ) {
+			$data['prompt_frequency'] = self::get_frequency_summary( $popup );
+			$data['prompt_placement'] = $popup['options']['placement'] ?? '';
+		}
+
+		$watched_blocks = [
+			'registration'             => 'newspack/reader-registration',
+			'donation'                 => 'newspack-blocks/donate',
+			'newsletters_subscription' => 'newspack-newsletters/subscribe',
+		];
+
+		$data['prompt_blocks'] = [];
+
+		foreach ( $watched_blocks as $key => $block_name ) {
+			if ( \has_block( $block_name, $popup['content'] ) ) {
+				$data['prompt_blocks'][] = $key;
+			}
+		}
+
+		$data['interaction_data'] = [];
+
+		return $data;
+	}
+
+	/**
+	 * Store the rendered popups data.
+	 *
+	 * @param array $popup The popup array representation.
+	 * @return void
+	 */
+	public static function get_rendered_popups( $popup ) {
+		$data = self::get_popup_metadata( $popup );
+		if ( ! empty( $data['prompt_id'] ) ) {
+			self::$popups[ $data['prompt_id'] ] = $data;
+		}
+	}
+
+	/**
+	 * Sanitizes the popup params to be sent as params for GA events
+	 *
+	 * All params in GA events must be strings, so we need to make the array flat and convert all values to strings.
+	 *
+	 * This method is also used by the Newspack Data Events API.
+	 *
+	 * @param array $popup_params The popup params as they are returned by Newspack_Popups_Data_Api::get_popup_metadata and by the prompt_interaction data.
+	 * @return array
+	 */
+	public static function prepare_popup_params_for_ga( $popup_params ) {
+		// Invalid input.
+		if ( ! is_array( $popup_params ) || ! isset( $popup_params['prompt_id'] ) ) {
+			return [];
+		}
+
+		$sanitized = $popup_params;
+
+		unset( $sanitized['interaction_data'] );
+		$sanitized = array_merge( $sanitized, $popup_params['interaction_data'] );
+
+		unset( $sanitized['prompt_blocks'] );
+		foreach ( $popup_params['prompt_blocks'] as $block ) {
+			$sanitized[ 'prompt_has_' . $block ] = 1;
+		}
+
+		// @TODO: How to handle prompts with more than one block?
```

Comment:
@leogermani If there's more than one block, what about making this a comma-delimited list? e.g. `donation, registration, newsletters_subscription`

This probably won't happen much in practice as our recommended/default prompts only contain a single type at a time.

Reply:
I'm not sure what data types GA supports for these values—if they can accept an array then that would probably be more useful.

Reply:
it only supports strings. up to 100 characters.

Yeah, we could concatenate them in alphabetical order I guess. But I wanted us to take a moment and see what are the cases this will happen to make sure this is the better way to go.

The only place I see this happening is in the new Memberships/content gating features - and we are not tracking membership purchases at all.. so I'd say that this is part of the task to add the Content gating data to the whole Data events api and ga4

Reply:
True, I guess I'm not seeing the utility of the `action_type` param right now.

> I wanted us to take a moment and see what are the cases this will happen to make sure this is the better way to go.

Membership-focused prompts are a good case. But then we may want to add additional `action_type` values to cover more membership-specific purposes, right? Custom publisher-authored prompts might also be another case, if they include both a donation and newsletter signup form in the same prompt (unlikely but could happen). I'm okay leaving this as-is for now until we have some concrete use case scenarios we need to support.

Reply:
> True, I guess I'm not seeing the utility of the action_type param right now.

Yeah, maybe it's redundand with the other we have.

I think we could discuss this and the `load` event subject with Katie to make a decision

Reply:
Since I was wrong about this, and the `load` GA3 event does only fire for prompts that will be displayed, I think this clarifies the usefulness of the `loaded` vs. `seen` events.